### PR TITLE
Log de-duplication and merging logs from multiple domains.

### PR DESCRIPTION
**Pull request by l-k-**
Deduplicate() method takes an Iterator and returns an Iterator with
duplicate facts removed from the stream.

Merge() method takes multiple Iterators, combines their fact streams
into one time-ordered stream, and and returns an Iterator for the
resulting stream.

### CODE REVIEW

**Comment by bruth** [comment]
Looks good. I suggest changing internal calls from `Err()` to simply referencing the `err` value. This is for aesthetic, but more importantly because some iterator implementations `Err()` hides [implementation details](https://github.com/chop-dbhi/origins/blob/master/csv.go#L290-L293) which could alter the break condition. For this reason, it would be best to set the convention to *not* use the public `Err()` for internal calls.

File: view/log.go [hunk 1]
```diff
@@ -140,6 +140,149 @@ func (li *logView) Err() error {
...
+
+// deduplicator maintains the state of a steam of unique facts
+type deduplicator struct {
+	iter          origins.Iterator
+	observedFacts *factMap
+	err           error
+}
+
+func (d *deduplicator) Err() error {
+	return d.err
```
Comment (bruth) [inline view/log.go]: This should return `d.iter.Err()`.

File: view/log.go [hunk 2]
```diff
@@ -140,6 +140,149 @@ func (li *logView) Err() error {
...
+
+func (d *deduplicator) Next() *origins.Fact {
+	if d.Err() != nil {
+		return nil
+	}
+
+	for {
+		f := d.iter.Next()
+
+		if d.err = d.iter.Err(); d.Err() != nil || f == nil {
```
Comment (bruth) [inline view/log.go]: There is no need to check for an error in the loop. If `d.iter` returns a nil value, the loop will break. The caller is expected to call `d.Err()` which proxies to `d.iter.Err()` (previous comment).

```go
func (d *deduplicator) Next() *origins.Fact {
    if d.iter.Err() != nil {
        return nil
    }

    var f *origins.Fact

    for {
        if f = d.iter.Next(); f == nil {
            break
        }

        if !d.observedFacts.add(f) {
            return f
        }
    }

    return nil
}
```

**Comment by l-k-** [comment]
Makes sense, thanks. Addressed.

**Pull request merged**